core/src/com/jediterm/terminal/model/LinesStorage.kt

```diff
@@ -0,0 +1,91 @@
+package com.jediterm.terminal.model
+
+import kotlin.math.min
+
+/**
+ * Base interface for storing terminal lines.
+ * For the first line of the storage, we use the term top, for the last line - bottom.
+ * Supports adding/removing lines from the top and bottom.
+ * Also accessing lines by index.
+ */
+internal interface LinesStorage : Iterable<TerminalLine> {
+  /** Count of the available lines in the storage */
+  val size: Int
+
+  /** [index] must be in the range [0, size) */
+  operator fun get(index: Int): TerminalLine
+
+  /** @return -1 if there is no such line */
+  fun indexOf(line: TerminalLine): Int
+
+  /**
+   * Adds a new line to the start of the storage.
+   * If the implementation limits the max capacity of the storage,
+   * and storage is full, then the line should not be added.
+   */
+  fun addToTop(line: TerminalLine)
+
+  /**
+   * Adds a new line to the end of the storage.
+   * If the implementation limits the max capacity of the storage,
+   * and storage is full, the first line from the top should be removed.
+   */
+  fun addToBottom(line: TerminalLine)
+
+  /**
+   * Removes a single line from the start of the storage.
+   * @throws NoSuchElementException if storage is empty.
+   */
+  fun removeFromTop(): TerminalLine
+
+  /**
+   * Removes a single line from the end of the storage.
+   * @throws NoSuchElementException if storage is empty.
+   */
+  fun removeFromBottom(): TerminalLine
+
+  fun clear()
+}
+
+internal fun LinesStorage.addAllToTop(lines: List<TerminalLine>) {
+  for (ind in lines.lastIndex downTo 0) {
+    addToTop(lines[ind])
+  }
+}
+
+internal fun LinesStorage.addAllToBottom(lines: List<TerminalLine>) {
+  for (line in lines) {
+    addToBottom(line)
+  }
+}
+
+internal fun LinesStorage.removeFromTop(count: Int): List<TerminalLine> {
+  if (count < 0) {
+    throw IllegalArgumentException("Count must be >= 0")
+  }
+  if (count == 0) {
+    return emptyList()
+  }
+  val countToRemove = min(count, size)
+  val result = ArrayList<TerminalLine>(countToRemove)
+  repeat(countToRemove) {
+    result.add(removeFromTop())
+  }
+  return result
+}
+
+internal fun LinesStorage.removeFromBottom(count: Int): List<TerminalLine> {
+  if (count < 0) {
+    throw IllegalArgumentException("Count must be >= 0")
+  }
+  if (count == 0) {
+    return emptyList()
+  }
+  val countToRemove = min(count, size)
+  val result = ArrayList<TerminalLine>(countToRemove)
+  repeat(countToRemove) {
+    result.add(removeFromBottom())
+  }
+  result.reverse()
+  return result
+}
```

Comment:
Would be nice to extract duplicate. Somethig like:

```
internal fun LinesStorage.removeFromTop(count: Int): List<TerminalLine> {
  return perform(count, false) {
    removeFromTop()
  }
}

internal fun LinesStorage.removeFromBottom(count: Int): List<TerminalLine> {
  return perform(count, true) {
    removeFromBottom()
  }
}

private inline fun LinesStorage.perform(count: Int, reverse: Boolean, operation: (() -> TerminalLine)): List<TerminalLine> {
  if (count < 0) {
    throw IllegalArgumentException("Count must be >= 0")
  }
  if (count == 0) {
    return emptyList()
  }
  val actualCount = min(count, size)
  if (actualCount == 1) {
    return listOf(operation())
  }
  val result = ArrayList<TerminalLine>(actualCount)
  repeat(actualCount) {
    result.add(operation())
  }
  if (reverse) {
    result.reverse()
  }
  return result
}
```

Not required, I can do it later myself.

Reply:
BTW, I like the idea of `addAllToTop`, `addAllToBottom`, `removeFromBottom` and `removeFromTop` as extension functions. Unlikely, it can be implemented much more efficient using `ArrayDeque` directly. Well, I guess a bit, but if it can be proven to have a clear advantage, then it can be converted to instance methods.

Reply:
I've seen the warning that there is a duplicated code. But failed to find a way how to extract common code preserving the readability. I think that duplication is OK here. Your suggestion is harder to understand to my taste.
But I don't mind if you prefer to refactor this place.

Reply:
I was also thinking about whether `addAll`/`removeAll` operations might be more effective when using `ArrayDeque` directly, but I think that benefit is negligible here.
If we prove that there is something significant, then OK, we can make instance methods. And I guess it is even should be binary compatible change.